File: GGAndroid_2.0/GGDemo/src/main/java/io/ganguo/demo/viewmodel/activity/HttpDemoVModel.java
package io.ganguo.demo.viewmodel.activity;

import android.view.View;

import java.util.concurrent.TimeUnit;

import io.ganguo.demo.module.impl.APIOneImpl;
import io.ganguo.demo.viewmodel.item.ItemOneListVModel;
import io.ganguo.demo.viewmodel.loading.OnListLoadingVModel;
import io.ganguo.library.ui.view.ActivityInterface;
import io.ganguo.rx.RxActions;
import io.ganguo.rx.RxCollections;
import io.ganguo.viewmodel.callback.ICenterLoadingView;
import io.ganguo.viewmodel.common.HFSRecyclerViewModel;
import io.ganguo.viewmodel.databinding.IncludeHfSwipeRecyclerBinding;
import io.ganguo.vmodel.rx.RxVMLifecycle;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.internal.functions.Functions;
import io.reactivex.schedulers.Schedulers;

/**
 * <p>
 * 网络请求Demo
 * </p>
 * Created by leo on 2018/7/30.
 */
public class HttpDemoVModel extends HFSRecyclerViewModel<ActivityInterface<IncludeHfSwipeRecyclerBinding>> {

    @Override
    public void onViewAttached(View view) {
        getOneList();

    }


    /**
     * function：获取一个列表数据
     */
    protected void getOneList() {
        APIOneImpl
                .get()
                .getOneList("3528")
                .subscribeOn(Schedulers.io())
                .delay(3000, TimeUnit.MILLISECONDS)
                .map(entity -> entity.getContentOneEntities())
                .compose(RxCollections.filterNotEmpty())
                .compose(RxCollections.emitItems())
                .map(e -> new ItemOneListVModel(e))
                .toList()
                .observeOn(AndroidSchedulers.mainThread())
                .map(itemOneListVModels -> itemOneListVModels)
                .doOnSuccess(itemOneListVModels -> {
                    getAdapter().clear();
                    getAdapter().addAll(itemOneListVModels);
                    getAdapter().notifyDiffUtilSetDataChanged();
                })
                .doFinally(() -> toggleEmptyView())
                .compose(RxVMLifecycle.bindSingleViewModel(this))
                .subscribe(Functions.emptyConsumer(), RxActions.printThrowable(getClass().getSimpleName() + "--getOneList--"));
    }


    @Override
    public void onRefresh() {
        super.onRefresh();
        getOneList();
    }

    @Override
    public void onLoadMore() {
        super.onLoadMore();
        getOneList();
    }


    @Override
    public ICenterLoadingView createLoadingView() {
        return new OnListLoadingVModel();
    }
}

File: GGAndroid_3.0/main/app-mvvm-v1-sample/src/main/java/io/ganguo/demo/viewmodel/activity/HttpDemoVModel.kt
package io.ganguo.demo.viewmodel.activity

import android.view.View
import com.scwang.smart.refresh.layout.api.RefreshLayout
import io.ganguo.demo.entity.one.MovieEntity
import io.ganguo.demo.http.api.impl.DouBanServiceImpl2
import io.ganguo.demo.http.helper.DouBanHttpHelper
import io.ganguo.demo.viewmodel.item.ItemOneListVModel
import io.ganguo.http2.error.Errors
import io.ganguo.viewmodel.core.viewinterface.ActivityInterface
import io.ganguo.log.core.Logger
import io.ganguo.rx.RxActions
import io.ganguo.rx.transformer.RxTransformer
import io.ganguo.rx.utils.emitItems
import io.ganguo.rx.utils.filterNotEmpty
import io.ganguo.viewmodel.pack.RxVMLifecycle
import io.ganguo.viewmodel.databinding.IncludeHfSwipeRecyclerBinding
import io.ganguo.viewmodel.pack.common.HFSRecyclerViewModel
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.internal.functions.Functions
import io.reactivex.rxkotlin.addTo
import io.reactivex.schedulers.Schedulers

/**
 * <pre>
 * author : leo
 * time   : 2018/11/01
 * desc   : 网络请求Demo
 * </pre>
 */
open class HttpDemoVModel : HFSRecyclerViewModel<ActivityInterface<IncludeHfSwipeRecyclerBinding>>() {


    override fun onViewAttached(view: View) {
        simulateErrorTransformer()
        getDouBanMovies()
        getGankBanners()
    }

    /**
     * 获取豆瓣开源Api数据
     */
    private fun getDouBanMovies() {
        DouBanServiceImpl2
                .get()
                .getDouBanMovies()
                .subscribeOn(Schedulers.io())
                .compose(RxTransformer.errorToastProcessor())
                .compose(DouBanHttpHelper.errorProcessor())
                .compose(DouBanHttpHelper.dataProcessor())
                .compose(filterNotEmpty())
                .compose<MovieEntity>(emitItems())
                .map { e -> ItemOneListVModel(e) }
                .toList()
                .observeOn(AndroidSchedulers.mainThread())
                .map { itemOneListVModels -> itemOneListVModels }
                .doOnSuccess { itemOneListVModels ->
                    adapter.clear()
                    adapter.addAll(itemOneListVModels)
                    adapter.notifySetDataDiffChanged()
                }
                .doFinally { toggleEmptyView() }
                .compose(RxVMLifecycle.bindSingleViewModel(this))
                .subscribe(Functions.emptyConsumer(), RxActions.printThrowable(javaClass.simpleName + "--getOneList--"))
                .addTo(lifecycleComposite)
    }


    /**
     * 获取Gank开源Banner api数据
     */
    private fun getGankBanners() {
        DouBanServiceImpl2
                .get()
                .getGankBanners()
                .subscribeOn(Schedulers.io())
                .compose(RxTransformer.errorToastProcessor())
                .compose(RxTransformer.errorLoggerProcessor())
                .observeOn(AndroidSchedulers.mainThread())
                .doOnNext {
                    Logger.e("getDataFromDynamicDomain：$it")
                }
                .subscribe(Functions.emptyConsumer(), RxActions.printThrowable(javaClass.simpleName + "--getDataFromDynamicDomain--"))
                .addTo(lifecycleComposite)
    }

    /**
     * 模仿错误异常处理
     */
    private fun simulateErrorTransformer() {
        Observable
                .error<Throwable>(Errors.AuthorizationException(0, "你的登录信息过期了"))
                .compose(RxTransformer.errorToastProcessor())
                .compose(RxTransformer.errorLoggerProcessor())
                .subscribe()
                .addTo(lifecycleComposite)
    }


    override fun onRefresh(refreshLayout: RefreshLayout) {
        super.onRefresh(refreshLayout)
        getDouBanMovies()
    }
}
